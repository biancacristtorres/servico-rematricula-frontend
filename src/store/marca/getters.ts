import { MarcaState } from './state';
import { GetterTree } from 'vuex';
import { RootState } from '..';
import { Instituicao } from '@/common/enum/Instituicao';
import { Marca } from '@/model/Marca';

export enum MarcaGetterTypes {
  ENCONTRAR_MARCA_POR = 'DISCIPLINA',
  OBTER_LINK_DO_SOL = 'OBTER_LINK_DO_SOL',
  E_SAO_JUDAS = 'E_SAO_JUDAS',
  E_UNI_BH = 'E_UNI_BH',
}

const getters: GetterTree<MarcaState, RootState> = {

  [MarcaGetterTypes.ENCONTRAR_MARCA_POR](state) {
    return (instituicao: Instituicao) => {
      const marcaEncontrada: Marca | undefined = state.marcas.find(
        (m) => m.possuiInstituicao(instituicao),
      );

      if (marcaEncontrada) {
        return marcaEncontrada;
      } else {
        throw new Error(`Nenhuma marca foi encontrada para a Instituição ${instituicao}`);
      }
    };
  },

  [MarcaGetterTypes.OBTER_LINK_DO_SOL](state, currentGetters) {
    return (instituicao: Instituicao) => {
      const marca: Marca = currentGetters[MarcaGetterTypes.ENCONTRAR_MARCA_POR](instituicao);
      return `${marca.linkBaseDoSOL}SOL/aluno/index.php/index/seguranca/dev/instituicao/${instituicao}`;
    };
  },

  [MarcaGetterTypes.E_SAO_JUDAS](state, currentGetters) {
    return (instituicao: Instituicao) => {
      const marca: Marca = currentGetters[MarcaGetterTypes.ENCONTRAR_MARCA_POR](instituicao);
      const ehSaoJudas = marca.nome === 'USJT';
      return ehSaoJudas;
    };
  },

  [MarcaGetterTypes.E_UNI_BH](state, currentGetters) {
    return (instituicao: Instituicao) => {
      const marca: Marca = currentGetters[MarcaGetterTypes.ENCONTRAR_MARCA_POR](instituicao);
      const ehUniBH = marca.nome === 'UNIBH';
      return ehUniBH;
    };
  },
};

export default getters;
