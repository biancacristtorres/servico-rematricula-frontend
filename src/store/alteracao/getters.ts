import { GetterTree } from 'vuex';
import { AlteracaoState } from './state';
import { RootState } from '@/store';
import { Disciplina } from '@/model/Disciplina';
import {
  FiltrarStateDeDisciplinaService,
} from '@/services/alteracao/FiltrarStateDeDisciplinaService';

export enum AlteracaoGetterTypes {
  DISCIPLINA = 'DISCIPLINA',
  POSSUI_TURMAS = 'POSSUI_TURMAS',
  TURMA_ATUAL = 'TURMA_ATUAL',
  ESTA_CARREGANDO = 'ESTA_CARREGANDO',
  POSSUI_ALTERACAO = 'POSSUI_ALTERACAO',
  FILTRO = 'FILTRO',
  TURMAS_FILTRADAS = 'TURMAS_FILTRADAS',
  TURMAS_EQUIVALENTES_FILTRADAS = 'TURMAS_EQUIVALENTES_FILTRADAS',
  TURMAS_ESPECIAIS_FILTRADAS = 'TURMAS_ESPECIAIS_FILTRADAS',
}

const filtrarStateDeDisciplinaService = new FiltrarStateDeDisciplinaService();

const getters: GetterTree<AlteracaoState, RootState> = {

  [AlteracaoGetterTypes.ESTA_CARREGANDO](state) {
    return state.alterando || state.preparandoTurmas;
  },

  [AlteracaoGetterTypes.DISCIPLINA](state) {
    const encontrarDisciplina = (codigoDaDisciplina: number) => filtrarStateDeDisciplinaService
      .processar(state.disciplinas, codigoDaDisciplina);

    return encontrarDisciplina;
  },

  [AlteracaoGetterTypes.POSSUI_TURMAS](state) {

    const possuiTurmas = (codigoDaDisciplina: number) => {
      const disciplina: Disciplina = filtrarStateDeDisciplinaService
        .processar(state.disciplinas, codigoDaDisciplina);

      return disciplina.turmas.length > 0 ||
        disciplina.turmasEquivalentes.length > 0 ||
        disciplina.turmasEspeciais.length > 0;
    };

    return possuiTurmas;
  },

  [AlteracaoGetterTypes.TURMA_ATUAL](state) {

    const turmaAtual = (codigoDaDisciplina: number) => {
      const disciplina: Disciplina = filtrarStateDeDisciplinaService
        .processar(state.disciplinas, codigoDaDisciplina);

      const atualDasTurmas = disciplina.turmas.find((t) => t.eTurmaAtual);
      const atualDasEquivalentes = disciplina.turmasEquivalentes.find((t) => t.eTurmaAtual);
      const atualDasEspeciais = disciplina.turmasEspeciais.find((t) => t.eTurmaAtual);

      return atualDasTurmas || atualDasEquivalentes || atualDasEspeciais;
    };

    return turmaAtual;
  },

  [AlteracaoGetterTypes.POSSUI_ALTERACAO](state) {
    return state.possuiAlteracao;
  },

  [AlteracaoGetterTypes.FILTRO](state) {
    return state.filtro;
  },

  [AlteracaoGetterTypes.TURMAS_FILTRADAS](state) {
    const turmaFiltrada = (codigoDaDisciplina: number) => {
      const disciplina: Disciplina = filtrarStateDeDisciplinaService.processar(
        state.disciplinas, codigoDaDisciplina,
      );

      return state.filtro.exibirTurnosDiferentes ?
        disciplina.turmas :
        disciplina.turmas.filter((t) => t.eDoTurnoDoAluno);
    };

    return turmaFiltrada;
  },

  [AlteracaoGetterTypes.TURMAS_EQUIVALENTES_FILTRADAS](state) {
    const turmasEquivalentesFiltradas = (codigoDaDisciplina: number) => {
      const disciplina: Disciplina = filtrarStateDeDisciplinaService.processar(
        state.disciplinas, codigoDaDisciplina,
      );

      return state.filtro.exibirTurnosDiferentes ?
        disciplina.turmasEquivalentes :
        disciplina.turmasEquivalentes.filter((t) => t.eDoTurnoDoAluno);
    };

    return turmasEquivalentesFiltradas;
  },

  [AlteracaoGetterTypes.TURMAS_ESPECIAIS_FILTRADAS](state) {
    const turmasEspeciaisFiltradas = (codigoDaDisciplina: number) => {
      const disciplina: Disciplina = filtrarStateDeDisciplinaService.processar(
        state.disciplinas, codigoDaDisciplina,
      );

      return state.filtro.exibirTurnosDiferentes ?
        disciplina.turmasEspeciais :
        disciplina.turmasEspeciais.filter((t) => t.eDoTurnoDoAluno);
    };

    return turmasEspeciaisFiltradas;
  },

};

export default getters;
