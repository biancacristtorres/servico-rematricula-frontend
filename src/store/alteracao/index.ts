import { Module } from 'vuex';
import { AlteracaoState, state } from './state';
import { RootState } from '@/store';
import mutations from './mutations';
import actions from './actions';
import getters from './getters';

const alteracao: Module<AlteracaoState, RootState> = {
  state,
  mutations,
  actions,
  getters,
  namespaced: true,
};

export default alteracao;
