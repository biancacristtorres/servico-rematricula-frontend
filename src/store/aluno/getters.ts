import { GetterTree } from 'vuex';
import { AlunoState } from './state';
import { RootState } from '@/store';

export enum AlunoGetterTypes {
  INFORMACOES_ALUNO = 'INFORMACOES_ALUNO',
  ESTA_OBTENDO_INFORMACOES_ALUNO = 'ESTA_OBTENDO_INFORMACOES_ALUNO',
}

const getters: GetterTree<AlunoState, RootState> = {

  [AlunoGetterTypes.INFORMACOES_ALUNO](state) {
    return state.informacoesAluno;
  },

  [AlunoGetterTypes.ESTA_OBTENDO_INFORMACOES_ALUNO](state) {
    return state.estaObtendoInformacoes;
  },

};

export default getters;
