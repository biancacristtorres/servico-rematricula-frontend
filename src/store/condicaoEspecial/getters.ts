import { GetterTree } from 'vuex';
import { CondicaoEspecialState } from './state';
import { RootState } from '@/store';
import { CondicaoEspecial } from '@/common/enum/CondicaoEspecial';

export enum GetterTypes {
  POSSUI_SOLICITADAS = 'POSSUI_SOLICITADAS',
  CONDICOES_ESPECIAIS = 'CONDICOES_ESPECIAIS',
}

const getters: GetterTree<CondicaoEspecialState, RootState> = {

  [GetterTypes.CONDICOES_ESPECIAIS](state) {
    return [
      CondicaoEspecial.TurmasEspecias,
      CondicaoEspecial.CampusDiferentes,
      CondicaoEspecial.TurnosDiferentes,
    ];
  },
};

export default getters;
