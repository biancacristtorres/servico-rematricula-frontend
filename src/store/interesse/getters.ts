import { GetterTree } from 'vuex';
import { InteresseState } from './state';
import { RootState } from '@/store';

export enum InteresseGetterTypes {
  DISCIPLINAS_PENDENTES = 'DISCIPLINAS_PENDENTES',
  DISCIPLINAS_ESTAO_CARREGADAS = 'DISCIPLINAS_ESTAO_CARREGADAS',
}

const getters: GetterTree<InteresseState, RootState> = {

  [InteresseGetterTypes.DISCIPLINAS_PENDENTES](state) {
    return state.disciplinas;
  },

};

export default getters;
