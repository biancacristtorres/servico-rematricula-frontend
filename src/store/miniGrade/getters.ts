import { GetterTree } from 'vuex';
import { MiniGradeState } from './state';
import { RootState } from '@/store';

export enum MiniGradeGetterTypes {
  MINI_GRADE = 'MINI_GRADE',
}

const getters: GetterTree<MiniGradeState, RootState> = {

  [MiniGradeGetterTypes.MINI_GRADE](state) {
    return state.horarios;
  },

};

export default getters;
