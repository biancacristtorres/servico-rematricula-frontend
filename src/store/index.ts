import Vue from 'vue';
import Vuex from 'vuex';
import autorizacao from 'componente-frontend-core/store/autorizacao';
import interesse from './interesse';
import condicaoEspecial from './condicaoEspecial';
import featureToggle from 'componente-frontend-core/store/featureToggle';
import alteracao from './alteracao';
import resumo from './resumo';
import aluno from './aluno';
import actions from './actions';
import marca from './marca';
import miniGrade from './miniGrade';
import { state } from './state';
import mutations from './mutations';

export { StoreNamespaces } from '@/store/namespaces';

Vue.use(Vuex);

export class RootState {
}

const modules = {
  autorizacao,
  interesse,
  condicaoEspecial,
  featureToggle,
  alteracao,
  resumo,
  aluno,
  marca,
  miniGrade,
};

const store = new Vuex.Store({
  modules,
  actions,
  state,
  mutations,
});

Object.assign(window, { $store: store });

export default store;
