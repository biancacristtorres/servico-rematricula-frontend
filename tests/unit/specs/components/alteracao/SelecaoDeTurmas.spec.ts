import {
  SelecaoDeTurmasFactory,
} from '../../../factory/components/alteracao/SelecaoDeTurmasFactory';
import store, { StoreNamespaces } from '@/store';
import { AlteracaoGetterTypes } from '@/store/alteracao/getters';

describe('Componente de Seleção de Turmas', () => {
  const factory = new SelecaoDeTurmasFactory();

  beforeEach(async (done) => {
    factory.dadoQueAStoreDeAlteracaoPossuaCalculoI();
    factory.prepararContainerDeInjecaoDaApi();
    factory.dadoQueONovoSelecionadorEstaDesabilitado();

    await factory.montarComponente();

    done();
  });

  afterEach(() => {
    factory.restaurarContainerDeInjecaoDaApi();
  });

  it('Carrega os dados de disciplina da store de alteração corretamente', () => {
    expect(factory.vueInstance.disciplina.codigo).toEqual(factory.codigoCalculoI);
  });

  it('Carrega os dados de turma da disciplina corretamente', () => {
    expect(factory.vueInstance.disciplina.turmas).toContainEqual(factory.turma);
  });

  it('Carrega os dados de turma equivalente da disciplina corretamente', () => {
    expect(factory.vueInstance.disciplina.turmasEquivalentes).toContainEqual(factory.turmaEquivalente);
  });

  it('Ao final da montagem, o componente está pronto para exibição', () => {
    expect(factory.vueInstance.prontoParaExibicao).toBeTruthy();
  });

  it('Escuta um evento de alteração de turma, faz um PUT e seleciona a turma', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerPUT();
    factory.dadoQueExistaUmAlunoNaStore();

    factory.turmaRenderizada().trigger('click');
    await factory.aguardarRenderizacao();

    const estaSelecionada = factory.estaSelecionada(factory.turmaRenderizada());
    expect(estaSelecionada).toBeTruthy();
    done();
  });

  it('Escuta um evento de alteração de turma equivalente, faz um PUT e seleciona a turma equivalente', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerPUT();
    factory.dadoQueExistaUmAlunoNaStore();

    factory.turmaEquivalenteRenderizada().trigger('click');
    await factory.aguardarRenderizacao();

    const estaSelecionada = factory.estaSelecionada(factory.turmaEquivalenteRenderizada());
    expect(estaSelecionada).toBeTruthy();
    done();
  });

  it('Ao alterar um turma, a mini grade é recarregada', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerPUT();
    factory.dadoQueExistaUmAlunoNaStore();

    factory.turmaRenderizada().trigger('click');
    await factory.aguardarRenderizacao();

    expect(factory.stateDaMiniGrade.horarios).toContainEqual(factory.horarioDaMiniGradeEsperado);
    done();
  });

  it('Escuta um evento de remoção de turma, faz um Delete e nenhuma opção é selecionada', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerDelete();

    const nenhumaOpcao = factory.componente.find(`#nenhuma-opcao-${ factory.codigoCalculoI }`);
    nenhumaOpcao.trigger('click');

    expect(factory.estaSelecionada(nenhumaOpcao)).toBeTruthy();
    done();
  });

  it('Ao remover um turma, a mini grade é recarregada', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerDelete();

    factory.nenhumaOpcaoRenderizada().trigger('click');
    await factory.aguardarRenderizacao();

    expect(factory.stateDaMiniGrade.horarios).toContainEqual(factory.horarioDaMiniGradeEsperado);
    done();
  });

  it('Quando um evento de remoção falhar, um feedback é exibido e nenhuma opção não é selecionada', async (done) => {
    factory.dadoQueApiRetorneErroAoFazerDelete();

    factory.nenhumaOpcaoRenderizada().trigger('click');

    await factory.aguardarRenderizacao();

    expect(factory.vueInstance.$dialogo.instance).not.toBeUndefined();
    expect(factory.estaSelecionada(factory.nenhumaOpcaoRenderizada())).toBeTruthy();
    done();
  });

  it('Alterna estado de loading ao selecionar uma turma', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerPUT();

    factory.turmaRenderizada().trigger('click');
    expect(factory.estaCarregando()).toBeTruthy();

    await factory.aguardarRenderizacao();
    expect(factory.estaCarregando()).toBeFalsy();
    done();
  });

  it('Alterna estado de loading ao remover uma turma', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerDelete();

    factory.nenhumaOpcaoRenderizada().trigger('click');
    expect(factory.estaCarregando()).toBeTruthy();

    await factory.aguardarRenderizacao();
    expect(factory.estaCarregando()).toBeFalsy();
    done();
  });

  it('Ao fazer uma seleção de turma, deve informar que houve alteração', async (done) => {
    factory.dadoQueApiRetorneStatusDeSucessoAoFazerPUT();

    factory.turmaRenderizada().trigger('click');

    const atual = store.getters[
      `${ StoreNamespaces.ALTERACAO }/${ AlteracaoGetterTypes.POSSUI_ALTERACAO }`
    ];
    expect(atual).toBeTruthy();
    done();
  });

  it('Snapshot', () => {
    expect(factory.componente).toMatchSnapshot();
  });
});
