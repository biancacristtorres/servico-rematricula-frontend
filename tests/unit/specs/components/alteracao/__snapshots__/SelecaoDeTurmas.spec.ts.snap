// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de Seleção de Turmas Snapshot 1`] = `
<div class="grey lighten-4">
  <div id="turmas-disciplina-1">
    <table>
      <tbody>
        <tr>
          <td class="td-radio text-xs-center"><i aria-hidden="true" id="nenhuma-opcao-1" value="true" class="v-icon v-icon--link material-icons theme--light primary--text" style="font-size: 16px;">

              radio_button_checked
            </i></td>
          <td><label for="nenhuma-opcao-1">
              Nenhuma das opções
            </label></td>
        </tr>
      </tbody>
    </table>
    <!---->
    <div class="cabecalho-turmas pa-3"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light">
        school
      </i> <span class="body-1">
    Turmas disponíveis
  </span></div>
    <div class="container-tabela" id="disciplina-turmas">
      <table id="disciplina-com-turmas" cellspacing="0" cellpadding="0">
        <thead id="turmas-cabecalho" class="primary white--text text-xs-center">
          <tr>
            <th></th>
            <!---->
            <th>Dia e Horário</th>
            <th>Campus</th>
            <th>Curso</th>
            <th>Turma</th>
            <th>Vagas</th>
            <!---->
          </tr>
        </thead>
        <tbody>
          <tr class="horario-turma text-xs-center">
            <td rowspan="1" class="td-radio"><i aria-hidden="true" id="turma-3" class="v-icon selecionar-turma v-icon--link material-icons theme--light primary--text" style="font-size: 16px;">

                radio_button_unchecked
              </i></td>
            <!---->
            <td>
              <li class="turmas-semanais">
                Quarta
              </li>
              <p class="ma-0">
                ()
              </p>
            </td>
            <td>
              Buritis
            </td>
            <td>
              CC
            </td>
            <td>
              X48
            </td>
            <td>
              10
            </td>
            <!---->
          </tr>
        </tbody>
      </table>
    </div>
    <div class="cabecalho-turmas pa-3"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light">
        school
      </i> <span class="body-1">
    Disciplinas equivalentes
  </span> <span class="v-tooltip v-tooltip--bottom"><div class="v-tooltip__content" style="left: 0px; max-width: 300px; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>
      Disciplinas equivalentes são lecionadas em outros cursos,
      porém contém a mesma base da disciplina que você deve cursar.
    </span></div><span><i aria-hidden="true" class="v-icon ml-3 material-icons theme--light" style="font-size: 16px;">
      help
    </i></span></span>
  </div>
  <div class="container-tabela" id="disciplina-turmas-equivalentes">
    <table id="disciplina-com-turmas" cellspacing="0" cellpadding="0">
      <thead id="turmas-cabecalho" class="primary white--text text-xs-center">
        <tr>
          <th></th>
          <th class="coluna-disciplina">
            Disciplina
          </th>
          <th>Dia e Horário</th>
          <th>Campus</th>
          <th>Curso</th>
          <th>Turma</th>
          <th>Vagas</th>
          <!---->
        </tr>
      </thead>
      <tbody>
        <tr class="horario-turma text-xs-center">
          <td rowspan="1" class="td-radio"><i aria-hidden="true" id="turma-1" class="v-icon selecionar-turma v-icon--link material-icons theme--light primary--text" style="font-size: 16px;">

              radio_button_unchecked
            </i></td>
          <td rowspan="1">

          </td>
          <td>
            <li class="turmas-semanais">
              Terça
            </li>
            <p class="ma-0">
              ()
            </p>
          </td>
          <td>
            Aimorés
          </td>
          <td>
            CC
          </td>
          <td>
            X40
          </td>
          <td>
            3
          </td>
          <!---->
        </tr>
      </tbody>
    </table>
  </div>
  <!---->
</div>
</div>
`;
